File: games/lynrummy/python/perf_harness.py
"""
perf_harness.py — load find_play snapshots, benchmark, profile.

Pipe `agent_game.py --capture FILE` to capture (hand, board)
inputs. Then run this harness against the JSONL file:

  - lists the slowest cases by recorded total_wall;
  - re-runs each top-N case `--repeats` times and reports
    median wall (independent of network jitter, since this
    only exercises the planner — no HTTP);
  - optionally cProfiles the slowest single case and dumps
    the top-30 cumulative-time entries.

Usage:
    python3 perf_harness.py /tmp/perf_snapshots.jsonl
    python3 perf_harness.py snaps.jsonl --top 5 --repeats 3
    python3 perf_harness.py snaps.jsonl --profile-slowest
"""

import argparse
import cProfile
import json
import pstats
import statistics
import sys
import time
from io import StringIO

import agent_prelude


def _to_tuple_card(c):
    """JSON stores cards as lists; the planner uses tuples."""
    return (c[0], c[1], c[2])


def _to_tuple_stack(s):
    return [_to_tuple_card(c) for c in s]


def _load_snapshots(path):
    out = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            rec = json.loads(line)
            rec["hand"] = [_to_tuple_card(c) for c in rec["hand"]]
            rec["board"] = [_to_tuple_stack(s) for s in rec["board"]]
            out.append(rec)
    return out


def _time_one(rec, repeats):
    """Re-run find_play `repeats` times, return list of walls."""
    walls = []
    for _ in range(repeats):
        t0 = time.time()
        agent_prelude.find_play(rec["hand"], rec["board"])
        walls.append(time.time() - t0)
    return walls


def _summarize(rec, walls):
    return {
        "captured_wall": rec["total_wall"],
        "median_wall": statistics.median(walls),
        "min_wall": min(walls),
        "max_wall": max(walls),
        "hand_size": len(rec["hand"]),
        "board_size": len(rec["board"]),
        "found_play": rec["found_play"],
        "n_projections": len(rec["projections"]),
    }


def _print_summary(rank, summary):
    print(f"  #{rank:2d} captured={summary['captured_wall']:5.2f}s "
          f"median={summary['median_wall']:5.2f}s "
          f"min={summary['min_wall']:5.2f}s "
          f"max={summary['max_wall']:5.2f}s | "
          f"hand={summary['hand_size']:>2} "
          f"board={summary['board_size']:>2} "
          f"projs={summary['n_projections']:>2} "
          f"{'+plan' if summary['found_play'] else 'STUCK'}")


def _print_projection_breakdown(rec):
    print("    projections:")
    for proj in rec["projections"]:
        cards = ",".join(f"{c[0]}/{c[1]}/{c[2]}" for c in proj["cards"])
        marker = "✓" if proj["found_plan"] else "·"
        print(f"      {marker} {proj['kind']:9} wall={proj['wall']:.2f}s "
              f"cards=[{cards}]")


def _profile_one(rec):
    profiler = cProfile.Profile()
    profiler.enable()
    agent_prelude.find_play(rec["hand"], rec["board"])
    profiler.disable()
    out = StringIO()
    stats = pstats.Stats(profiler, stream=out)
    stats.sort_stats("cumulative")
    stats.print_stats(30)
    return out.getvalue()


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("snapshots",
                    help="JSONL file from agent_game --capture")
    ap.add_argument("--top", type=int, default=10,
                    help="Show the N slowest cases (default 10)")
    ap.add_argument("--repeats", type=int, default=5,
                    help="Re-time each case N times (default 5)")
    ap.add_argument("--profile-slowest", action="store_true",
                    help="cProfile the slowest case")
    args = ap.parse_args()

    snaps = _load_snapshots(args.snapshots)
    if not snaps:
        sys.exit(f"no snapshots in {args.snapshots}")

    snaps.sort(key=lambda s: s["total_wall"], reverse=True)
    top = snaps[:args.top]

    print(f"Loaded {len(snaps)} snapshots; profiling top {len(top)} "
          f"with {args.repeats} repeats each.\n")
    print("Per-case re-times:")

    summaries = []
    for i, rec in enumerate(top, 1):
        walls = _time_one(rec, args.repeats)
        s = _summarize(rec, walls)
        summaries.append(s)
        _print_summary(i, s)

    print("\nProjection breakdown for the slowest case:")
    _print_projection_breakdown(top[0])

    if args.profile_slowest:
        print("\ncProfile (top 30 by cumulative time) for the "
              "slowest case:")
        print(_profile_one(top[0]))


if __name__ == "__main__":
    main()

File: games/lynrummy/ts/bench/perf_harness.ts
// perf_harness.ts — load find_play snapshots, benchmark.
//
// Reads a JSONL file of captured `find_play` snapshots and:
//   - lists the slowest cases by recorded total_wall;
//   - re-runs each top-N case --repeats times and reports median wall
//     (independent of network jitter — only exercises the planner).
//
// For deeper profiling: `node --prof bench/perf_harness.ts ...` then
// `node --prof-process` on the log.
//
// NOTE: the original snapshot-capture mechanism was Python-side and
// retired with the Python subtree. Until a TS-side capture lands,
// this harness has no input source; kept against the day it's needed.
//
// Usage:
//   node bench/perf_harness.ts /tmp/perf_snapshots.jsonl
//   node bench/perf_harness.ts snaps.jsonl --top=5 --repeats=3
//   node bench/perf_harness.ts snaps.jsonl --max-states=10000

import * as fs from "node:fs";

import type { Card } from "../src/rules/card.ts";
import { findPlay, type PlayStats } from "../src/hand_play.ts";

interface CapturedProjection {
  kind: "pair" | "singleton";
  cards: number[][];
  wall: number;
  found_plan: boolean;
}

interface CapturedSnapshot {
  hand: number[][];
  board: number[][][];
  projections: CapturedProjection[];
  total_wall: number;
  found_play: boolean;
}

function asCard(c: number[]): Card {
  return [c[0]!, c[1]!, c[2]!] as const;
}

function loadSnapshots(p: string): CapturedSnapshot[] {
  const out: CapturedSnapshot[] = [];
  const text = fs.readFileSync(p, "utf8");
  for (const raw of text.split("\n")) {
    const line = raw.trim();
    if (!line) continue;
    out.push(JSON.parse(line));
  }
  return out;
}

function median(xs: readonly number[]): number {
  const a = [...xs].sort((x, y) => x - y);
  const n = a.length;
  if (n === 0) return 0;
  return n % 2 === 1 ? a[(n - 1) >> 1]! : (a[n / 2 - 1]! + a[n / 2]!) / 2;
}

interface Summary {
  capturedWall: number;
  medianWall: number;
  minWall: number;
  maxWall: number;
  handSize: number;
  boardSize: number;
  foundPlay: boolean;
  nProjections: number;
}

function timeOne(rec: CapturedSnapshot, repeats: number, maxStates: number): { walls: number[]; lastStats: PlayStats } {
  const hand = rec.hand.map(asCard);
  const board = rec.board.map(s => s.map(asCard));
  let lastStats: PlayStats = { totalWallMs: 0, projections: [] };
  const walls: number[] = [];
  for (let i = 0; i < repeats; i++) {
    lastStats = { totalWallMs: 0, projections: [] };
    const t0 = performance.now();
    findPlay(hand, board, { maxStates, stats: lastStats });
    walls.push((performance.now() - t0) / 1000); // seconds, to match Python
  }
  return { walls, lastStats };
}

function summarize(rec: CapturedSnapshot, walls: number[], _stats: PlayStats): Summary {
  return {
    capturedWall: rec.total_wall,
    medianWall: median(walls),
    minWall: Math.min(...walls),
    maxWall: Math.max(...walls),
    handSize: rec.hand.length,
    boardSize: rec.board.length,
    foundPlay: rec.found_play,
    nProjections: rec.projections.length,
  };
}

function printSummary(rank: number, s: Summary): void {
  console.log(
    `  #${String(rank).padStart(2, " ")} captured=${s.capturedWall.toFixed(2).padStart(5)}s ` +
      `median=${s.medianWall.toFixed(2).padStart(5)}s ` +
      `min=${s.minWall.toFixed(2).padStart(5)}s ` +
      `max=${s.maxWall.toFixed(2).padStart(5)}s | ` +
      `hand=${String(s.handSize).padStart(2)} ` +
      `board=${String(s.boardSize).padStart(2)} ` +
      `projs=${String(s.nProjections).padStart(2)} ` +
      `${s.foundPlay ? "+plan" : "STUCK"}`,
  );
}

function printProjectionBreakdown(rec: CapturedSnapshot): void {
  console.log("    projections:");
  for (const proj of rec.projections) {
    const cards = proj.cards.map(c => `${c[0]}/${c[1]}/${c[2]}`).join(",");
    const marker = proj.found_plan ? "✓" : "·";
    console.log(`      ${marker} ${proj.kind.padEnd(9)} wall=${proj.wall.toFixed(2)}s cards=[${cards}]`);
  }
}

interface Args {
  snapshots: string;
  top: number;
  repeats: number;
  maxStates: number;
  maxCapturedWall: number;
}

function parseArgs(argv: string[]): Args {
  let snapshots = "";
  let top = 10;
  let repeats = 5;
  let maxStates = 10000;
  let maxCapturedWall = 30.0;
  for (const a of argv) {
    if (a.startsWith("--top=")) top = parseInt(a.slice("--top=".length), 10);
    else if (a.startsWith("--repeats=")) repeats = parseInt(a.slice("--repeats=".length), 10);
    else if (a.startsWith("--max-states=")) maxStates = parseInt(a.slice("--max-states=".length), 10);
    else if (a.startsWith("--max-captured-wall=")) maxCapturedWall = parseFloat(a.slice("--max-captured-wall=".length));
    else if (!a.startsWith("--")) snapshots = a;
  }
  if (!snapshots) {
    process.stderr.write("usage: node bench/perf_harness.ts <snapshots.jsonl> [--top=N --repeats=N --max-states=N]\n");
    process.exit(2);
  }
  return { snapshots, top, repeats, maxStates, maxCapturedWall };
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  let snaps = loadSnapshots(args.snapshots);
  if (snaps.length === 0) {
    process.stderr.write(`no snapshots in ${args.snapshots}\n`);
    process.exit(1);
  }
  snaps = snaps.filter(s => s.total_wall <= args.maxCapturedWall);
  snaps.sort((a, b) => b.total_wall - a.total_wall);
  const top = snaps.slice(0, args.top);

  console.log(
    `Loaded ${snaps.length} snapshots (post-filter); profiling top ${top.length} with ${args.repeats} repeats each, max_states=${args.maxStates}.\n`,
  );
  console.log("Per-case re-times:");

  const summaries: Summary[] = [];
  for (let i = 0; i < top.length; i++) {
    const { walls, lastStats } = timeOne(top[i]!, args.repeats, args.maxStates);
    const s = summarize(top[i]!, walls, lastStats);
    summaries.push(s);
    printSummary(i + 1, s);
  }

  if (top.length > 0) {
    console.log("\nProjection breakdown for the slowest case:");
    printProjectionBreakdown(top[0]!);
  }
}

main();
